### DISCUSSION TITLE
Adding an `boa_interop` crate for interoperability with the interpreter.

### DISCUSSION THREAD

**Discussion by hansl** [boa-dev/boa#3770] [Ideas]
A bit of context; I'm working on a firmware on an embedded system and I want to use Boa to add JavaScript capabilities with the internal functions (think DB, UI, Filesystem, that sort of thing). I have some synthetic modules implemented in my R&D branch, but find the boiler plate for interacting with the script a little too much for my taste.

What I'd like to see is something akin to (in Rust, and not working yet):

```rust
use boa_engine::{Context, JsResult, Module, Trace};
use boa_interop::{js_module, JsRest};

#[js_module]
trait MyModule {
  // Simplest form of a callable function.
  fn add(&self, arg1: u32) -> u32;

  // Another form where arguments are passed as rest, and has a context.
  fn sum(&self, args: &JsRest, context: &Context) -> JsResult<JsValue>;

  // A JavaScript getter.
  #[getter]
  fn start(&self) -> u32;
}

#[derive(Clone, Trace)
struct MyModuleImpl {
  start: u32
}

impl MyModuleImpl {
  pub fn new(start: u32) -> Self {
    Self { start }
  }
}

impl MyModule for MyModuleImpl {
  fn add(&self, arg1: u32) -> u32 {
    self.start + arg1
  }

  fn sum(&self, args: &JsRest, context: &Context) -> JsResult<JsValue> {
    let mut total = self.start;
    for a in args {
      let a: u32 = a.try_into_js(ctx)?;
      total += a;
    }

    Ok(total)
  }

  fn start(&self) -> u32 {
    self.start
  }
}

pub fn create_module(context: &mut Context) -> JsResult<Module> {
  MyModuleImpl::new(123).into_synthetic_module(context)
}
```

Then from JavaScript:

```js
let module = import(/* module loading dependent */ "my_module");
console.log(module.add(1)); // 124
console.log(module.sum(1, 2, 3, 4, 5)); // 138
console.log(module.start); // 123
```

I think this would make creating and registering synthetic modules much much easier.

Unless this effort is already underway or there is strong opposition for this, I volunteer to add the new crate and do the work to build that API.

Let me know if there's interest.

**Comment by jedel1043** [comment]
That would be pretty cool! We're definitely interested on seeing better integration APIs for embedders, akin to wasm-bindgen. If you have the time to do the work, I'd definitely be open to offering some guidance on it.

The crate would preferably be developed incrementally, so that you don't need to open a big PR with all the changes; this makes the review work a lot easier and it would also let us participate in the design process at the start, when it's a lot more important.

**Comment by hansl** [comment]
I started https://github.com/boa-dev/boa/pull/3772 with a first step. I'll wait for a review of it before continuing this effort. Thanks!
